docs/http-status-codes.adoc

```diff
@@ -0,0 +1,166 @@
+|Check for alerts or other known failures (e.g., sleds offline, disks that need replacement).  **NOTE:** it would be a bug if the operator resolved all the issues reported by the system and they're still getting 503 errors.
+|Check server logs, metrics, etc. for details and debug
+
+|===
+
+== Rationale
+
+People often observe that HTTP is woefully underspecified.  One manifestation is that people choose different status codes for the same conditions.  Choosing a status code for a particular condition is hard because many of the defined status codes overlap or aren't super clear:
+
+* A successful response that needs no response content could reasonably return a "200 OK" or a "204 No Content".
+* A request that successfully creates a resource could reasonably return "200 OK" or "201 Created".
+* If a request is unauthorized, both "403 Forbidden" or "404 Not Found" could apply, depending on whether the server's willing to tell the client that the resource exists and they just don't have access to it vs. act like it doesn't exist because the client doesn't have permissions to know that it exists.
+* If a request has no credentials, a "401 Unauthorized" or "403 Forbidden" might both seem to apply.  The spec is not super clear on this.
+
+To make things more confusing: some codes are badly named (e.g., "401 Unauthorized" reflects a problem with the _authentication credentials_).  And there's not even a complete list of codes to begin with: many commonly-used status codes come not from one of the main HTTP specs but some other related spec (like WebDAV).
+
+Clients are expected to treat any unrecognized code as the corresponding "x00" code, which makes those a catch-all that's often a reasonable choice.
+
+**While the choice of HTTP status code for a particular condition can be arbitrary, that's not the same as saying it doesn't matter!**  A consistent and thoughtful set of choices can make an HTTP-based service significantly easier for both users and the people operating the service.
+
+When we choose what status code to use for a condition, we should consider:
+
+* what the spec says (which is often insufficient for the reasons mentioned above)
+* what users are likely to expect (for better or worse, other popular APIs and StackOverflow threads are useful data points here)
+* whether a particular status code (or distinction between codes) conveys useful information to anybody.  See <<_how_people_respond_to_status_codes>> above.
+
+The rest of this section describes the non-obvious choices we've made about status codes.
+
+=== Success status codes
+
+==== "200 OK" vs. "204 No Content"
+
+If an endpoint may ever return content, we use "200 OK" always (even if the response body is zero bytes sometimes).  This is simpler on both the server and client sides.
+
+Some sources suggest using "204 No Content" for PUT endpoints, which is a reasonable choice.  We usually return "200 OK" with the new representation of the resource, which seems more useful.
```

Comment:
I only hedged out of laziness -- I hadn't audited all existing PUT endpoints.  I've done that now and will make this stronger because there are no exceptions today.

Reply:
Fixed in 01a7cb36.